File: practice/gild/netconnections/submitc4/netconnections.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
/* Disjoint Set */
int *set = 0;
int f(int n) /* Set Find */
{
   if(set[n] < 0) return n;
   else return set[n] = f(set[n]);
}
void u(int n1, int n2) /* Set Union */
{
   n1 = f(n1); n2 = f(n2);
   if(set[n2] < set[n1]) set[n1] = n2;
   else {
      if(set[n1] == set[n2]) set[n1]--;
      set[n2] = n1;
   }
}
int main(int argc, char **argv)
{
   char c; int i, num, p1, p2, y = 0, n = 0;
   FILE *fp = fopen(argv[1], "r");
   fscanf(fp, "%d", &num);
   set = malloc((num+1) * sizeof(int));
   for(i = 0; i<num+1; i++) set[i] = -1;
   while(!feof(fp))
   {
      if(!fscanf(fp, "%c%d%d", &c, &p1, &p2)) break;
      switch(c) {
         case 'c': if(f(p1) != f(p2)) u(p1, p2); break;
         case 'q': (f(p1) == f(p2)) ? y++ : n++; break;
      }
   }
   printf("%d,%d\n", y, n);
   fclose(fp); free(set); 
   return 0;
}
      
  

File: practice/gild/netconnections/submitpy2/netconnections.py
import sys, array

file = open(sys.argv[1])
num = int(next(file))
arr = array.array('i', [-1] * num)

def find(n):
  if(arr[n] < 0):
    return n
  else:
    t = find(arr[n])
    arr[n] = t
    return arr[n]

def union(n1, n2):
  n1 = find(n1)
  n2 = find(n2)
  if(arr[n2] < arr[n1]):
    arr[n1] = n2
  else:
    if(arr[n1] == arr[n2]):
      arr[n1] = arr[n1]-1
    arr[n2] = n1


y = n = 0
for line in file:
  sp = line.split()
  c = sp[0]
  p1 = int(sp[1])
  p2 = int(sp[2])

  if c == 'c':
    if find(p1) != find(p2):
      union(p1, p2)
  else:
    if find(p1) == find(p2):
      y += 1
    else:
      n += 1

print ("%s,%s") % (y,n)
